nwplus/nwhacks2019
# Web/internal/filter

:tickets: **Ticket(s)**: Closes #227 

---

## :construction_worker: Changes

This implements the filter modal (all of it except for "filter top n" functionality). It already adds about 800 lines of code so I thought it might be better to push this out first to make the review process less painful.

You can:
- filter by a number of properties, as well as tags
- you can also enter custom fields and values

![image](https://i.imgur.com/LdnZZAe.gif)

## :thought_balloon: Notes

General implementation:
- After filters are applied, a filter expression is generated (which is a function that takes an applicant as input and returns `true` if the applicant passes the combined filter).
  - This filter expression is applied to a flattened version of each applicant object in `applicants`. This means that if the applicant has some nested value like `"tags": {"someTag": true}`, it will be flattened into `"tags_someTag": "true"`. Note that all primitive values in a flattened applicant are converted into a string, but this is fine because JavaScript does type conversion when applying the filters (with non-strict operators like `==` and `!=`).
- I've commented the functions that I've added to make it easier to digest, let me know if something needs clarifying.

Regrets:
- The UI logic took a while to implement and ended up being a ton of code even after refactoring. Might have been better to push out a simple "Enter filter expression" text box to save time but oh well lol

## :flashlight: Testing Instructions
- Try applying filters. We don't have any applicants with tags yet, but you can manually add tags to some applicant to test it out.
- Check that the Apply button enables/disables at the right time
- Check that the Cancel button reverts any changes to the filters
- Try anything you want to try and break the UI state / filters
